Support for regex flags

maximkulkin opened the pull request:
Implement support for regex flags
    
This change adds support for re.IGNORECASE, re.DOTALL and
re.UNICODE regex flags. Flags are extracted from compiled
regex. Also, it supports subpattern flags introduced in
Python 3.6.
    
re.IGNORECASE controls literals: if it is enabled, instead
of generating static literal, it generates either
lowercase or uppercase character.
    
re.DOTALL: if enabled, "." matches newlines.
    
re.UNICODE controls character categories (e.g. "\d" or
"\w"): if enabled, categories include unicode characters.
In Python 2 default is to generate ASCII only data, in
Python 3 - unicode. Python 3 also has reverse flag -
re.ASCII to disable unicode character generation.

Fixes #2, #3 and #5

maximkulkin committed:
Improve tests checking invariants for regex strategy

maximkulkin committed:
Implement support for regex flags

This change adds support for re.IGNORECASE, re.DOTALL and
re.UNICODE regex flags. Flags are extracted from compiled
regex. Also, it supports subpattern flags introduced in
Python 3.6.

re.IGNORECASE controls literals: if it is enabled, instead
of generating static literal, it generates either
lowercase or uppercase character.

re.DOTALL: if enabled, "." matches newlines.

re.UNICODE controls character categories (e.g. "\d" or
"\w"): if enabled, categories include unicode characters.
In Python 2 default is to generate ASCII only data, in
Python 3 - unicode. Python 3 also has reverse flag -
re.ASCII to disable unicode character generation.

maximkulkin commented:
@Zac-HD Could you take a look at this PR. It is full of humiliation and suffering (thanks to Unicode)

Zac-HD commented on tests/test_regex.py:
```diff
@@ -1,37 +1,267 @@
+                )
+
+    def test_matching_ascii_word_chars(self):
+        self._test_matching_pattern('\\w', is_word)
```
Prefer `r'\w'` to `'\\w'`, especially for regex.

Zac-HD commented on tests/test_regex.py:
```diff
@@ -1,37 +1,267 @@
+
+    h.find(strategy, lambda s: compiled_pattern.match(s) is not None)
+
+    with pytest.raises(he.NoSuchExample):
```
This should be a separate test so it could fail independently.

maximkulkin replied:
Well, this is the main purpose of the assert: check that every generated string built from given regex matches regex (or in other words: "there are no generated examples that do not match regex"). I will actually remove the previous positive case (the find() above) to make it more clear, I think that find() is redundant.

Zac-HD commented on tests/test_regex.py:
```diff
@@ -1,37 +1,267 @@
+
+    with pytest.raises(he.NoSuchExample):
+        failed_example = h.find(strategy, lambda s: compiled_pattern.match(s) is None)
+        assert False, 'Failed example: %r' % failed_example
```
A better message would be something like `'found %r using strategy %s, which does not match the pattern'` - the source of the example is important here.

Zac-HD commented on tests/test_regex.py:
```diff
@@ -1,37 +1,267 @@
         assert_can_generate('abc')
+        assert_can_generate('[a][b][c]')
+
+    def test_literals_with_ignorecase(self):
```
Instead of repeating this, use `pytest.mark.parameterize` or similar to apply the same test to each regex separately.  Same for `test_not_literal_with_ignorecase`.

Zac-HD commented on tests/test_regex.py:
```diff
@@ -1,37 +1,267 @@
+class TestRegexUnicodeMatching:
+    def _test_matching_pattern(self, pattern, isvalidchar, unicode=False):
+        if unicode:
+            r = re.compile(pattern) if six.PY3 else re.compile(pattern, re.UNICODE)
```
You can unconditionally use `re.compile(pattern, re.UNICODE)` - the flag doesn't do anything under Py3, but still exists.  Unfortunately `re.ASCII` is new, but you use it enough that I'd suggest adding a `re_compile_ASCII(pattern)` function to avoid the inline conditionals.

maximkulkin replied:
Although I see no harm in inline conditionals, I will still introduce functions to compile ascii/unicode regexes just to make code more compact.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,268 @@
+
+        if len(strategies) == 0:
+            return hs.just(u'')
+        elif len(strategies) == 1:
```
`hs.one_of` does this internally when passed a single strategy, so no need to replicate the special case here.

maximkulkin replied:
Thanks for pointing that out

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,268 @@
-def _strategy(codes, cache):
+
+def _strategy(codes, context):
     if not isinstance(codes, tuple):
```
It took me a few cycles through to understand the logic here - it would be nice to clarify this with an explanatory comment, eg 

    # codes is a recursive structure of lists where leaves are (code, value) tuples.
    # We handle the recursion here, and the base case further down.

Then I would also dedent the base-case and add a comment (eg `# base case handled here`) at the top of that.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,268 @@
+        elif len(strategies) == 1:
+            return strategies[0]
+        else:
+            return hs.tuples(*strategies).map(u''.join)
```
Same comments as for `CharacterBuilder.strategy` above.

maximkulkin replied:
This is a bit different and `hs.tuples()` behavior is different from `hs.one_of()`: empty tuples are still tuples, one element tuple is a tuple, not the element itself. So it is still beneficial to differentiate those 3 cases.

Strategies can be empty (e.g. empty group) and this should not invalidate the whole generation attempt, so generating empty string is an OK variant.

Zac-HD replied:
`hs.tuples(*strategies).map(u''.join)` works correctly for length-zero and length-one tuples - `''.join(()) == ''` and `''.join(('a',)) == 'a'` - but this *is* slightly more efficient and makes it easier to lift literals later.

maximkulkin replied:
Do I understand it correctly that `tuples()` is less efficient?

Zac-HD replied:
It is less efficient, but by a small enough margin that I would use it anyway for clarity.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,268 @@
+                    )
+                )
+
+        if len(strategies) == 0:
```
This makes me suspicious.  Can it ever happen?  If so, is the empty string really correct output?  (serious question)  If not, just `assert strategies` to make that clear and avoid hiding the problem.

maximkulkin replied:
Well, regular expression itself can't have an empty range (e.g. it can't be '[]'), but this is a separate builder object and might be used differently, so it should still maintain invariant: it returns a valid strategy that generates whatever you configured.

Zac-HD reviewed (commented):
Looks good to me.  The design is nice and makes sense, so specific comments inline are mostly about the tests.

Merging adjacent literals is certainly a good start on #5, but I think more is possible - so don't close it just yet!

maximkulkin commented:
Addressed all PR comments (unless otherwise stated), added a lot more documentation and code comments.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,268 @@
+        elif len(strategies) == 1:
+            return strategies[0]
+        else:
+            return hs.tuples(*strategies).map(u''.join)
```
`hs.tuples(*strategies).map(u''.join)` works correctly for length-zero and length-one tuples - `''.join(()) == ''` and `''.join(('a',)) == 'a'` - but this *is* slightly more efficient and makes it easier to lift literals later.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,325 @@
+
         elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
             return hs.just('')
```
Raise `InvalidArgument` and explain why the given regex cannot be used with this strategy - otherwise output will not match the pattern used to generate it, and that's an invariant we don't want to break.

maximkulkin replied:
It doesn't mean that it's not working. Usually people place "^" at the beginning of their regexes and "$" - at the end. And in that case it will work just fine. I also think about cases where "^" or "$" are in the middle and MULTILINE is enabled, so that those chars can generate newlines.

Also, "\b" (word boundary) and "\B" (not a word boundary) might work.

To enforce those invariants, I'm considering of adding a filter to resulting strategy that will check that generated example actually matches regex. What do you think about that?

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,325 @@
+        [
+            ('literal', 97),
+            ('literal', 98),
+            ('max_repeat', (1, 4294967295L, [
```
Drop the trailing `L` - code examples should use Python 3.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,325 @@
+            if HAS_SUBPATTERN_FLAGS:
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = hs.tuples(*[_strategy(part, context)
```
This looks like a flattened version of the return from CharacterBuilder.strategy - is it worth splitting this up too for the same reasons?

Zac-HD reviewed (commented):
Generally looking good - I particularly like the documentation.  Minor comments apart from "don't break that invariant".

The bigger one is actually in testing.  You have a common pattern of asserting that you can find an example that matches some predicate, and can't find one that matches the negation.  However `find` is the wrong tool for the first part of this - you should use `@given` and assert that *every* example matches the predicate, not that you can look for an example without errors. 

    # instead of `h.find(strategy, predicate)`, 
    @given(strategy)
    def inner(ex): 
        assert predicate(ex)
    inner()

You may want a separate helper function for this, and certainly should integrate it into `assert_all_examples`.

maximkulkin commented:
Ok, I didn't know about this `given` trick (although it doesn't look elegant), so I used the reverse: using `find()` to search for examples that do not match. Obviously, those two variants can be used interchangeably. I added an extra "positive check" (the one that you're concerned about) just to be safe.

I will look into `given` trick.

maximkulkin commented:
New update, changed the way tests are run (using `given`). Also added filter by matching generated examples with real regex. Added support for "groupref_exists" contruct ("(?(id)yes-pattern|no-pattern)") and a smarter generation for "$".

Zac-HD commented on tests/test_regex.py:
```diff
@@ -74,5 +300,37 @@ def test_named_groups(self):
+        r = re.compile('a^b')
+
+        with pytest.raises(he.NoSuchExample):
+            h.find(regex(r), r.match)
```
Does this not generate anything at all, in which case you can get the exception from `regex(r).example()`?  Or does it not generate any matches, in which case it should detect that and raise `InvalidArgument`?

This is the everything-matches invariant again - it's fine to not support some valid regexes, but they should raise InvalidArgument instead of exploding later.

maximkulkin replied:
You probably out of context here. Thing is that this regex "a^b" should not match anything. You can try that. But the way generator works currently it would generate strings "ab", which is incorrect. So I added this test and update generation to prevent generation in that case.

Moreover, figuring out that some regex would not match anything (and thus it should raise InvalidArgument) is very complicated task and I would like not to do that. Instead I think it is reasonable to just add filter that tries to use regular expression engine to see if generated examples match.

Zac-HD replied:
1. `'a^b'` does not match any string.  Thanks to the filter, the strategy therefore generates no examples.  (so far, so good)
2. Using `r.match` as the predicate is therefore redundant.  
3. `lambda s: True` would be more efficient, and clearly does not exclude possible outputs.
4. Using `regex(r).example()` is more efficient again - it basically uses `find` internally and stops at the first example, where `find` keeps trying until it finds the simplest example.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,335 @@
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            if value == sre.AT_END:
+                return hs.one_of(hs.just(u''), hs.just(u'\n'))
             return hs.just('')
```
This makes me nervous, but I can easily be convinced that it's correct with the regex-pattern-strategy I posted on HypothesisWorks/hypothesis-python#662.

Zac-HD commented on hypothesis_regex.py:
```diff
@@ -1,102 +1,335 @@
+
+    codes = sre.parse(pattern)
+
+    return _strategy(codes, Context(flags=flags)).filter(regex.match)
```
I would strongly prefer to avoid filtering, and instead have tests that convince us it's not needed - using a filter can easily hide more subtle problems that should be fixed instead, and it's a smallish but needless inefficiency.

maximkulkin replied:
There are some cases, which are either hard to ensure are generated correctly, nor to restructure to maintain invariants. E.g. ".*(?<!foo)bar$" (match anything that ends with "bar" unless "bar" is preceeded with "foo"). It is hard to accomplish that without using `filter()` somewhere in strategy (you could say "hey, let's get whatever strategies preceed that and then filter", but is almost the same as filtering the whole expression although whole expression filter is easier to implement).

My preference is to do best effort generating stuff without putting complex logic of rejecting regexes that in some cases we can not reliably generate, and let people and/or health checks to judge. At the end users can rewrite their regexes to make them less complex.

Zac-HD replied:
Fair enough - it's just a matter of differing preferences then, and this is your project!

I may try to add the rejection logic when we upstream this, but won't mention it again here.

Zac-HD reviewed (commented):
Smaller comments again - mostly that I don't like filtering, and found a quick-and-dirty way to generate patterns (by... filtering, I did say it was dirty).

The group-reference handling is lovely, and I can feel this pull getting closer 😄 .  Once it's merged, I think the biggest changes involved in upstreaming will be switching over the 2/3 compatibility layers!

Zac-HD reviewed (approved):
This looks really good.  :clinking_glasses:  My remaining comments are either nitpicking or well beyond the scope of this pull, so I'll leave it here unless you request specific feedback.
